### ISSUE TITLE
[RFC]: Lazy CUDA Graph capture

### ISSUE THREAD

**Issue by lionelvillard** [vllm-project/vllm#20098]
### Motivation.

Currently vLLM captures cudagraphs as part of the engine initialization significantly slowing down vLLM startup time. By default, vLLM captures 67 graphs, which depending on model size and GPU type, can take more than 10s. This is not great UX (see #19824 for details).

In addition, It's most unlikely that all 67 graphs are actually needed, wasting both time and space.  

### Proposed Change.

We propose to capture cudagraphs lazily. Instead of performing dummy runs during the engine initialization phase, the idea is to do those runs somewhere in the CUDA piecewise backend, and only for the current runtime shape if not cached already.

Exact implementation needs to be worked out.

### Feedback Period.

one week

### CC List.

@ProExpertProg @aarnphm @charlesfrye  

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ProExpertProg** [comment]
Cc @youkaichao @zou3519

**Comment by ProExpertProg** [comment]
A few thoughts:
- I think the current implementation in CudaPiecewiseBackend already kind-of supports lazy capture. It's just always initiated from the model runner during `capture_model`. 
- We should decide the interface for this, and make it robust. Whether there are separate `eager_cudagraph_sizes` and `lazy_cudagraph_sizes`, or if lazy is a subset of all, we should make sure that 1. specifying this on the CLI is intuitive and 2. that the config is the single source of truth for this.
- We could also try to add a feature to the scheduler where it lazily captures cudagraphs if there are no requests to be scheduled. But I'm not sure how common that is in the use-cases we care about (especially auto-scaling)

**Comment by zou3519** [comment]
A mode to do lazy cudagraphs capture seems reasonable. I'm curious if we think this should be the default or not. Doing cudagraphs lazily means that model serving time becomes less predictable.

**Comment by charlesfrye** [comment]
Nice! [We](https://modal.com/) and our users run a lot of vLLM services that autoscale, including scaling from 0, so startup time matters to us. When scaling from 0, the latency has to be incurred anyway, so lazy capture helps us spread the load out without compromising on later inferences.

Some details on that below!

## When would you care about capturing CUDA graphs but not want to do it eagerly?

IME, graph capture is most important for 8B and smaller models. Intuitively: CUDA kernel launch overhead is ~microsecond scale per kernel, so you're talking _maybe_ an ~millisecond improvement in ITL, which matters most when your ITL is also sub-10ms. I see roughly those latencies for models with ~4-8 GB of weights on short contexts on 1xH100 with vLLM/SGLang/TRT-LLM, cf [these benchmark results](https://modal.com/llm-almanac/advisor?metric=itl&filters=model%3DLlama+3.1+8B+fp8%2Ctokens%3D128%3B1024). Speed-of-light here for ITL is also 2-5ms.

Because the model weights are smaller, these are also the engines that boot fastest. In our system weights are loaded from storage at ~GB/s. I think that's typical for deployments of "moderate" sophistication, because you'd need IB/RocE and a weight server get to TB/s. So that part of the engine load is at the seconds scale, whereas it's at the minutes scale for larger models.

In this setting, we are sensitive to the extra latency of CUAD graph capture and the ITL benefits. Boots are ~45s without it, so ~10s would add 20%, cf [this PR/write-up](https://github.com/modal-labs/modal-examples/pull/1217).

## Other comments

> We should decide the interface for this, and make it robust. Whether there are separate eager_cudagraph_sizes and lazy_cudagraph_sizes, or if lazy is a subset of all, we should make sure that 1. specifying this on the CLI is intuitive and that the config is the single source of truth for this.

+1 for separate eager and lazy CUDA graph sizes. I expect we'd want to capture one or two graphs eagerly (say, batch size one and max) and leave the other, less common, sizes to lazy capture.

>  I'm curious if we think this should be the default or not. Doing cudagraphs lazily means that model serving time becomes less predictable.

IME developing and iterating on a lot of shorter-lived vLLM instances, a bit of extra variability is a very appealing trade for faster startup. The most important use case I can think of where this is also true is for RL off of rollouts -- training code is often short-lived and involves a lot of iteration by researchers. Once you finalize your training code for a run, you might still leave lazy capture in place. Tail latencies are less important, but throughput (which depends on average ITL and TTFT) matters.

**Comment by BoyuanFeng** [comment]
One concern is the performance drop after start up. By current eager CUDAGraph capture, everything happens during startup time. We can say for sure it will give consistently good performance during runtime. But for lazy CUDAGraph capture, we will see abrupt performance drop during runtime and it would be hard to know when the performance will be stable. Many people prefer consistent performance during runtime over slightly shorter start up time.

Comparing with lazy cudagraph capture, probably reducing number of cudagraph capture sizes is better, especially if we know what batch sizes we care about. Also, full-model cudagraph instead of piecewise cudagraph can also reduce graph capture size. For piecewise cudagraph, we are recording #layers x #sizes CUDAGraphs, which could be ~5000 CUDAGraphs for LLAMA-3.1-70B.

Question: What would be the batch size you care about? Currently we record 67 sizes, including [8, 16, 24, 32, ..., 512] + [1,2,4].

**Comment by yeqcharlotte** [comment]
> We propose to capture cudagraphs lazily. Instead of performing dummy runs during the engine initialization phase, the idea is to do those runs somewhere in the CUDA piecewise backend, and only for the current runtime shape if not cached already.

Another downside of doing this is CUDA graph takes substantial amount of memory that couldn't be taken into account in memory profiling. We'd rather see OOMs instantly and instruct users to start with a different gpu memory utilization than seeing unpredicted OOMs in the middle of some running processes when workload changes. 

I feel finding a good balance around how much to capture would make sense @BoyuanFeng said. Say maybe we can have different modes depends on compile optimization modes, say allowing batch size intervals to be uniform, exponential, or more sparse. 

I would also imagine diminishing returns over certain batch sizes. I think we should do some benchmarking around it.

Running batch sizes is really workload specific. Typical chatbot workload for TP8, TP1 sized models of ~2K input typically fall within batch size 16 - 64 in order to satisfy the TPOT requirements.

**Comment by aarnphm** [comment]
> Say maybe we can have different modes depends on compile optimization modes, say allowing batch size intervals to be uniform, exponential, or more sparse.

This sounds a bit too complex. I felt like given that this is workload specific, a guide on informing users what bs works for certain usage (function calling, rag, long context) would be a great starter.
+1 for reducing the batch size. For smaller models, we can potentially enable full cuda graph, whereas with larger models (70B and up) we can start with capped batch size to maximum of 128 would be good enough.

**Comment by lionelvillard** [comment]
> Another downside of doing this is CUDA graph takes substantial amount of memory that couldn't be taken into account in memory profiling. 

right but that can be solved by caching the max GPU memory utilization to use for a given configuration, similar to caching pytorch compilation artifacts. At a high level it seems this could work.

This makes me think we could also cache the result of peak memory usage profiling.

**Comment by angkywilliam** [comment]
One additional use case I’d like to add is related to LoRA. I’ve noticed that initiating vLLM with a higher `max_loras` value (36 in my case) significantly increases the startup time to as high as **7 minutes for a 8B model**. My hypothesis is that vLLM creates CUDA graph for each LoRA configuration, which leads to a large number of graphs being generated.

It would be ideal if we could combine prebuilding a subset of CUDA graphs with lazy loading. I believe this would be helpful for both merged and LoRA model use cases, allowing commonly used CUDA graphs to be prebuilt while less common configurations to be built on the fly. For the LoRA case in specific, it would be ideal to prebuilt graphs for up to, say, 8 LoRAs and have the remaining LoRA built on the fly as it's a less common traffic pattern.

**Comment by qthequartermasterman** [comment]
I think having prompt embeddings enabled vs disabled is relevant to this conversation. They're currently only supported in v0 (haven't yet been ported to v1, I'm hoping to do that before the next release 👀 ), but in the v0 engine, they're currently disabled by default partly because it doubled the cuda graph capture time (the graph had to be compiled separately for prompt token id inputs and prompt embedding inputs).  Depending on how the CUDA graph compilation occurs in v1, doing this lazily may have consequences for how prompt embeddings should be enabled.

**Comment by LucasWilkinson** [comment]
> Another downside of doing this is CUDA graph takes substantial amount of memory that couldn't be taken into account in memory profiling. We'd rather see OOMs instantly and instruct users to start with a different gpu memory utilization than seeing unpredicted OOMs in the middle of some running processes when workload changes.

+1 to this; I think this is a major concern. But if it's off by default and we can document the increased potential for OOMs (and as a result recommend much lower gpu memory utilization) then it could be a nice option for some users / development.

>right but that can be solved by caching the max GPU memory utilization to use for a given configuration, similar to caching pytorch compilation artifacts. At a high level it seems this could work.

Im not sure if this solves the issue since the cudagraphs that would get built are workload dependent so its really hard to say what the cache represents; not to mention this would mean that the user would have to warm-up the cache using a calibration workload if they wanted to deploy with lower OOM risk. Another option would be to get the size of 1 cudagraph and try to estimate the size of all possible ones with a safety factor


tl; dr Overall I think OOM issue is a big one but if we leave leave this feature off by default and document that this should only be used in environments where a higher risk of OOM is acceptable (i.e. development / autoscaling groups with good recovery) it could be a nice option

**Comment by lionelvillard** [comment]
Capturing CUDA graphs is done after memory profiling so there are already not taken into account. 

Lazy graph capture does not mean capturing more graphs which, by default, it's 67. It's still 67 max.

**Comment by LucasWilkinson** [comment]
> Capturing CUDA graphs is done after memory profiling so there are already not taken into account.

Sorry I should have been more clear; Im mainly refer to the expectations once the server is stood-up (i.e. ready to receive requests; after memory profiling, cuda-graph capture etc.). The chances of an OOM after that point is very low (ideally zero but we are not quite there). I believe this is what @yeqcharlotte was referring to. This doesnt really have anything really to do with the memory profiling but instead the idea that if we are going to raise OOM we'd ideally raise OOM when the server is standing-up instead of randomly when its up and running. This allows the user to quickly change the `gpu_memory_utilization` and try again. 

Lazy cuda-graphs would change this behavior, while you are probably correct that if `gpu_memory_utilization` is set to the same value with and without lazy cuda-graphs enabled the likelyhood we would raise OOM with lazy cuda-graphs but not with it disabled is pretty low, i.e. echoing: 

> Lazy graph capture does not mean capturing more graphs which, by default, it's 67. It's still 67 max.

I think there is still alot of cases where users tweak `gpu_memory_utilization`. With lazy cuda-graphs they could accidentally set this too high after being lured into a false sense of security since the server stood-up and a few test batches pass fine; due to the historical precedent that if the server stood-up your `gpu_memory_utilization` is likely good. Only to be surprised a few hours/days later if theres a burst of traffic or usually large prefills that trigger new cuda-graph captures.

This is why I was saying I think this should be an opt-in feature with appropriate documentation around how to add appropriate buffer the `gpu_memory_utilization` (i.e. set it much lower than the point where the server OOMs during start-up; much being relative to the historical precedent).

I think we could make it opt-out if we come up with sufficiently robust way to warn the user during start-up if there is a enough memory for all of the lazy cuda-graphs but that feels like it could be done as follow-up. This is what I was getting at with:

> Another option would be to get the size of 1 cudagraph and try to estimate the size of all possible ones with a safety factor

Hope that clears it up!

**Comment by lionelvillard** [comment]
You made it crystal clear, thanks @LucasWilkinson!

Thanks to @mgoin CUDA graph capture optimization, capturing 1 graph is now much cheaper (~50ms for a 8B model). Instead of lazily capturing graphs, we could interleave capturing graphs and doing a real forward passes. The system would become much more predicable.

**Comment by ProExpertProg** [comment]
> we could interleave capturing graphs and doing a real forward passes. The system would become much more predicable.

Sorry, what would this help? I think CUDAGraph capture is now so fast we don't need to address it further especially if it would increase complexity

**Comment by lionelvillard** [comment]
it's much faster than before but it is still in the order of seconds, which for model-as-a-service type of scenarios, it's still too long. 

We'll see about the added complexity. We should open a PR soon where we can evaluate if it is worthwhile or not.

**Comment by youkaichao** [comment]
I have a strong no for this feature. capturing cudagraph during runtime is very complicated, with huge maintenance burden, and it can break many other features.

If you want to reduce the start-up time, reduce the number of shapes you want to capture. We can work on providing a better ui to enable it. Right now it's already possible with cli arg, but maybe not very convenient.

**Comment by diegocastanibm** [comment]
Thank you all for the comments about this PR. We consider that this feature is important in the context of LLM distribution, such as LLM-d, that strive to deliver a distributed LLM inference framework with autoscaling features. A critical factor in this context is the reduction of startup time to allow for a much more efficient routing and scheduling. The CUDA Graph (CG) capture process consumes significant time during engine initialization, notably slowing down vLLM's startup speed. 
By default, vLLM captures 67 graphs, a process that can extend startup time beyond 10 seconds, depending on model size and GPU type. Some of your comments suggest limiting the number of CG’s during the initialization. However, we have noticed that the batch size to determine the CG can be quite unpredictable as can be seen in Fig. 1 below. The figure shows 4 runs of the benchmark using 1000 prompts each.

<img width="672" height="903" alt="Image" src="https://github.com/user-attachments/assets/b331797f-8e8b-4e56-989b-c88fa31c1dcc" />

Fig1: Batch sizes for CUDA graph in 1000 prompts running the benchmark 4 times


Fig. 2 shows the same experiment but clustering the 4 histograms in the same plot with different colors for each run. The multi-bar histogram clearly shows that the batch size can be very different and that it is not easy to anticipate the most suitable CG beforehand. Therefore, we consider that reducing the sizes of the CG can affect the runtime performance dramatically.

<img width="1306" height="633" alt="Image" src="https://github.com/user-attachments/assets/18495905-1847-450f-b7b1-3ffd7e01904c" />

Figure 2: Multibar histogram for 4 runs of benchmark with 1000 prompts each


To address this issue, we propose the approach in the PR: delayed CUDAgraph capture. Rather than conducting CUDAGraph capture during the engine initialization phase, we suggest executing these tasks during runtime at the very beginning. As a result, the CG capture will be time-bound. Luka suggested that we can indicate with a flag if we want to do it in a delayed fashion, keeping the CUDA graph capture during initialization as default. Also, to avoid the OOM problem, we will initialize the vLLM capturing the CG with the largest size.
Our dual objectives with this strategy are to minimize startup time in a 'cold start' scenario and manage the initial unpredictability of the inference process. Recent experimental results have demonstrated substantial reductions in startup time, particularly for larger models, while maintaining minimal impact on metrics such as TimeToFirstToken (TTFT).
We strongly believe that this feature can help LLM-d (and other tools for LLM distributions) making the scheduling process more efficient.

cc. @lionelvillard @youkaichao @ProExpertProg @WoosukKwon @ywang96 @LucasWilkinson @mgoin @simon-mo
